runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -1381,6 +1381,92 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
             break;
             }
          }
+      case TR::java_lang_Class_isHidden:
+      case TR::java_lang_Class_isAnonymousClass:
+      case TR::java_lang_Class_isEnum:
+      case TR::java_lang_Class_isSynthetic:
+      case TR::java_lang_Class_isArray:
+      case TR::java_lang_Class_isPrimitive:
+         {
+         // Only constrain the call in the last run of vp to avoid adding the candidate twice if the call is inside a loop
+         if (!lastTimeThrough())
+            return;
+         TR::Node *jlClassChild = node->getFirstChild();
+         bool jlClassChildGlobal;
+         TR::VPConstraint *jlClassChildConstraint = getConstraint(jlClassChild, jlClassChildGlobal);
+         int32_t result = -1;
+         if (  jlClassChildConstraint
+            && jlClassChildConstraint->isFixedClass()
+            && jlClassChildConstraint->getKnownObject())
+            {
+            if (rm == TR::java_lang_Class_isHidden)
+               result = comp()->fej9()->isHiddenClass(jlClassChildConstraint->getClass());
+            else if (rm == TR::java_lang_Class_isAnonymousClass)
+               result = comp()->fej9()->isAnonymousClass(jlClassChildConstraint->getClass());
+            else if (rm == TR::java_lang_Class_isEnum)
+               result = comp()->fej9()->isEnumClass(jlClassChildConstraint->getClass());
+            else if (rm == TR::java_lang_Class_isSynthetic)
+               result = comp()->fej9()->isSyntheticClass(jlClassChildConstraint->getClass());
+            else if (rm == TR::java_lang_Class_isArray)
+               result = comp()->fej9()->isPrimitiveArray(jlClassChildConstraint->getClass());
```

Comment:
Is this the right semantic if folding is done ? i.e. would we want to fold to a constant `false` if we had a reference array in this case ?

Reply:
Oh, I missed the second part of the check here, which would be `|| comp()->fej9()->isReferenceArray(jlClassChildConstraint->getClass());`. I will fix that.